File: test/cpp/test_anonymize_subject_info.cpp
#include <iostream>
#include <string>
#include <cstdio>
#include <fstream>
#include <vector>

#include "../src/harmonie_reader.h"

using namespace Harmonie;

static void printSubject(const char *label, const HarmonieReader::SUBJECT_INFO &info) {
    std::cout << "--- " << label << " ---" << std::endl;
    std::cout << "  id: " << info.id << std::endl;
    std::cout << "  firstname: " << info.firstname << std::endl;
    std::cout << "  lastname: " << info.lastname << std::endl;
    std::cout << "  sex: " << info.sex << std::endl;
    std::cout << "  birthDate: " << info.birthDate << std::endl;
    std::cout << "  age: " << info.age << std::endl;
    std::cout << "  height: " << info.height << std::endl;
    std::cout << "  weight: " << info.weight << std::endl;
}

static bool copyFile(const std::string &src, const std::string &dst) {
    std::ifstream in(src, std::ios::binary);
    if (!in) {
        return false;
    }
    std::ofstream out(dst, std::ios::binary);
    if (!out) {
        return false;
    }
    out << in.rdbuf();
    return static_cast<bool>(out);
}

int main(int argc, char **argv) {
    HarmonieReader reader;
    std::cout << "Test: test_anonymize_subject_info" << std::endl;

    std::string filename = "";
    std::string replacementId = "ANON";
    // Or set a path here when debugging from the IDE:
    // filename = "C:\\path\\to\\recording.sts";

    if (argc >= 2) {
        filename = argv[1];
    }
    if (argc >= 3) {
        replacementId = argv[2];
    }

    if (filename.empty()) {
        std::cout << "ERROR: No filename specified" << std::endl;
        std::cout << "Usage: test_anonymize_subject_info <file.sts> [replacement_id]" << std::endl;
        return 1;
    }

    std::string anonFile = filename;
    size_t dot = anonFile.find_last_of('.');
    if (dot == std::string::npos) {
        anonFile += "_anon";
    } else {
        anonFile = anonFile.substr(0, dot) + "_anon" + anonFile.substr(dot);
    }

    std::cout << "Copying " << filename << " -> " << anonFile << std::endl;
    if (!copyFile(filename, anonFile)) {
        std::cout << "ERROR: Could not copy file" << std::endl;
        return 1;
    }

    std::cout << "Opening copy: " << anonFile << std::endl;
    if (!reader.openFile(anonFile)) {
        std::cout << "ERROR Failed to open the file:" << anonFile << std::endl;
        std::cout << reader.getLastError() << std::endl;
        return 1;
    }

    HarmonieReader::SUBJECT_INFO before = reader.getSubjectInfo();
    printSubject("BEFORE", before);

    std::cout << "Anonymizing with replacement_id=" << replacementId << " ..." << std::endl;
    if (!reader.anonymizeSubjectInfo(replacementId, true)) {
        std::cout << "ERROR anonymize failed: " << reader.getLastError() << std::endl;
        return 1;
    }

    std::cout << "Saving file..." << std::endl;
    if (!reader.saveFile()) {
        std::cout << "ERROR saveFile failed" << std::endl;
        return 1;
    }
    reader.closeFile();

    std::string bak = anonFile + ".bak";
    if (std::remove(bak.c_str()) == 0) {
        std::cout << "Removed backup with original PHI: " << bak << std::endl;
    }

    std::cout << "Reopening anonymized file for verification..." << std::endl;
    HarmonieReader validation;
    if (!validation.openFile(anonFile)) {
        std::cout << "ERROR Failed to reopen:" << anonFile << std::endl;
        return 1;
    }

    HarmonieReader::SUBJECT_INFO after = validation.getSubjectInfo();
    printSubject("AFTER", after);
    validation.closeFile();

    bool ok = (after.id == replacementId) &&
              (after.firstname == "ANON") &&
              (after.lastname == "ANON") &&
              (after.birthDate == 0) &&
              (after.height == 0) &&
              (after.weight == 0);

    if (!ok) {
        std::cout << "ERROR anonymization incomplete" << std::endl;
        return 1;
    }

    std::cout << "SUCCESS Subject info anonymized. Output file: " << anonFile << std::endl;
    std::cout << "DONE" << std::endl;
    return 0;
}

File: test/python/test_anonymize_subject_info.py

"""
Anonymize Harmonie .sts patient/header metadata, then verify by reopening.

Usage (from repo root, after Release build of the Python module):
  python test/python/test_anonymize_subject_info.py path\\to\\recording.sts
  python test/python/test_anonymize_subject_info.py path\\to\\recording.sts SUBJ001

Works on a copy (*_anon.sts) so the original is left unchanged.
Also copies the companion .sig/.SIG if present.
Deletes the .bak created by save_file().
"""

import os
import shutil
import sys
from sys import platform

sys.path.insert(0, "")

if platform == "linux" or platform == "linux2":
    import build.HarmonieReader as HarmonieReader
elif platform == "darwin":
    import build.HarmonieReader as HarmonieReader
elif platform == "win32":
    import build.Release.HarmonieReader as HarmonieReader


def find_sig(sts_path):
    base, _ = os.path.splitext(sts_path)
    for ext in (".sig", ".SIG"):
        candidate = base + ext
        if os.path.isfile(candidate):
            return candidate
    return None


def print_subject(label, info):
    print(f"--- {label} ---")
    print(f"  id:         {info.id}")
    print(f"  firstname:  {info.firstname}")
    print(f"  lastname:   {info.lastname}")
    print(f"  sex:        {info.sex}")
    print(f"  birth_date: {info.birth_date}")
    print(f"  age:        {info.age}")
    print(f"  height:     {info.height}")
    print(f"  weight:     {info.weight}")


def is_anonymized(info, replacement_id):
    checks = [
        (info.id == replacement_id, "id"),
        (info.firstname == "ANON", "firstname"),
        (info.lastname == "ANON", "lastname"),
        (info.birth_date == 0, "birth_date"),
        (info.height == 0, "height"),
        (info.weight == 0, "weight"),
    ]
    failed = [name for ok, name in checks if not ok]
    return failed


def main():
    print("Test: test_anonymize_subject_info")

    filename = ""
    replacement_id = "ANON"
    if len(sys.argv) >= 2:
        filename = sys.argv[1]
    if len(sys.argv) >= 3:
        replacement_id = sys.argv[2]

    # Or set a path here when running from the IDE:
    # filename = r"C:\path\to\recording.sts"

    if filename == "":
        print("ERROR: No filename specified")
        print("Usage: python test/python/test_anonymize_subject_info.py <file.sts> [replacement_id]")
        quit(1)

    if not os.path.isfile(filename):
        print(f"ERROR: File not found: {filename}")
        quit(1)

    base, ext = os.path.splitext(filename)
    anon_sts = base + "_anon" + ext

    print(f"Copying {filename} -> {anon_sts}")
    shutil.copy2(filename, anon_sts)

    sig_src = find_sig(filename)
    if sig_src:
        anon_sig = base + "_anon" + os.path.splitext(sig_src)[1]
        print(f"Copying companion signal {sig_src} -> {anon_sig}")
        shutil.copy2(sig_src, anon_sig)
    else:
        print("WARNING: No companion .sig/.SIG found (signal read not required for this test)")

    reader = HarmonieReader.HarmonieReader()
    print(f"Opening copy: {anon_sts}")
    if not reader.open_file(anon_sts):
        print(f"ERROR Failed to open: {anon_sts}")
        print(reader.get_last_error())
        quit(1)

    before = reader.get_subject_info()
    print_subject("BEFORE", before)
    original_lastname = before.lastname
    original_firstname = before.firstname

    print(f"Anonymizing with replacement_id={replacement_id} ...")
    if not reader.anonymize_subject_info(replacement_id, True):
        print(f"ERROR anonymize failed: {reader.get_last_error()}")
        quit(1)

    print("Saving file...")
    if not reader.save_file():
        print("ERROR save_file failed")
        quit(1)
    reader.close_file()

    bak = anon_sts + ".bak"
    if os.path.isfile(bak):
        print(f"Removing backup with original PHI: {bak}")
        os.remove(bak)

    print("Reopening anonymized file for verification...")
    validation = HarmonieReader.HarmonieReader()
    if not validation.open_file(anon_sts):
        print(f"ERROR Failed to reopen: {anon_sts}")
        print(validation.get_last_error())
        quit(1)

    after = validation.get_subject_info()
    print_subject("AFTER", after)
    validation.close_file()

    failed = is_anonymized(after, replacement_id)
    if failed:
        print(f"ERROR anonymization incomplete for fields: {failed}")
        quit(1)

    # Best-effort binary search: original names should not remain in .sts
    if original_lastname and original_lastname not in ("ANON",):
        with open(anon_sts, "rb") as f:
            data = f.read()
        # Harmonie stores Latin-1 text; try common encodings
        for name in (original_lastname, original_firstname):
            if not name or name == "ANON":
                continue
            for encoding in ("latin-1", "utf-8"):
                try:
                    needle = name.encode(encoding)
                except UnicodeEncodeError:
                    continue
                if needle in data:
                    print(f"WARNING: original string still present in file bytes: {name!r} ({encoding})")

    print(f"SUCCESS Subject info anonymized. Output file: {anon_sts}")
    print("DONE")


if __name__ == "__main__":
    main()
